# Alignement, Read Groups, Mark duplicates {#ch:align}

Once the *fastq* files are ready to be processed we can align them with *STAR*. [Subread/Rsubread](http://subread.sourceforge.net/) is another widely used RNA-Seq aligner. The reason why I initially choose *STAR* over *Subread* was simply due to the fact that *STAR* can generate specific output for chimeric reads that can be directly used with [STAR-Fusion](https://github.com/STAR-Fusion/STAR-Fusion/wiki) to analyse gene fusions (see more in Section \@(ch:fusions)). Also, *STAR* is suggested in the the [GATK Best Practices to call variants in RNA-Seq](https://gatkforums.broadinstitute.org/gatk/discussion/3892/the-gatk-best-practices-for-variant-calling-on-rnaseq-in-full-detail). 

## Create `STAR` index 

STAR requires to build an index for the reference genome that will be used in the analignment and fusion calling step.

```{bash eval=FALSE}
# Iniaitlise Genome directory where to save STAR Index and STAR Fusion Index folders
star_genome100=path_to_genome_directory/star_index_hg19_99
mkdir -p ${star_genome100}
```


To build the *STAR index* one needs to provide the FASTA file for the reference genome used, a GTF file with information aabout the annotation and STAR also require an an extra parameter called `sjdbOverhang` which is usually set to be *(read length - 1)*. See `STAR` documentation for **Generating genome indexes** in the [STAR manual](https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf)
- 99 is (read length - 1) relative to the samples that I was working with.

Below is a wrapper for `STAR` call to build an index. 

```{bash eval=FALSE}
outpud_dir=../results
../functions/build_STAR_index.sh $genome_fasta_path $gtf_path $outpud_dir "hg19" 99
```


## STAR-1pass

If you are working with a cohort of bamfiles, STAR developer suggests to run the alignment in a two-pass mode. This consists of first aligning all the bamfiles, collecting the *splice junctions* as output of STAR and realign all the bamfiles with this new information. For more details about 1-pass, 2-pass-multi and 2-pass-single see Section 8 of the [STAR documentation](https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf). In my pipeline I normally use the 2-pass multi strategy as below.

```{bash eval=FALSE, echo=FALSE}
# save list of fastq files into a variable
samplein_runs=$(find ${fastqdir} -maxdepth 1 -name "*1.fastq*" | sort)

for fastqin in ${samplein_runs}; do 

  FQ1=$fastqin
  FQ2=${FQ1/_1/_2}
  
  bamout=$(echo $(basename $fastqin) | cut -f1 -d "_")
  
  lock=${star_1pass}/${bamout}.lock
  
  if [[ -d ${lock} ]] ; then 
  
    echo "${bamout} is being processed"
    
  continue
    
  else
    
    mkdir -p ${lock}
    
    Rscript ${workdir}/run_STAR.R --genome_index ${star_index} --fastqfiles $FQ1,$FQ2 \
    --sampleName ${bamout} --outdir ${star_1pass} --STARmode "1Pass"
    
    rmdir ${lock}

  fi

done

```


```{bash eval=FALSE, echo=FALSE}
# If HPC is available:
# save list of fastq files into a variable
samplein_runs=$(find ${fastqdir} -maxdepth 1 -name "*1.fastq*" | sort)

for fastqin in $samplein_runs ; do

  FQ1=$fastqin
  FQ2=${FQ1/_1/_2}
  
  bamout=$(echo $(basename $fastqin) | cut -f1 -d "_") # to be personalised according to file name

  echo '#!/bin/bash' > ${star_1pass}/${bamout}_align.sh
  echo "#PBS -q submit" >> ${star_1pass}/${bamout}_align.sh
  echo "#PBS -l nodes=1:ppn=1,mem=1gb" >> ${star_1pass}/${bamout}_align.sh
  echo "#PBS -N ${bamout}" >> ${star_1pass}/${bamout}_align.sh
  echo "#PBS -o ${bamout}_out" >> ${star_1pass}/${bamout}_align.sh
  echo "#PBS -e ${bamout}_err" >> ${star_1pass}/${bamout}_align.sh
  echo  "" >> ${star_1pass}/${bamout}_align.sh
  echo 'module load STAR' >> ${star_1pass}/${bamout}_align.sh
  echo 'module load R' >> ${star_1pass}/${bamout}_align.sh
  
  echo Rscript ${workdir}/run_STAR.R --genome_index ${star_genome100} --fastqfiles $FQ1,$FQ2 \
    --sampleName ${bamout} --outdir ${star_1pass} --STARmode "1Pass" >> ${star_1pass}/${bamout}_align.sh

done


for align in ${star_1pass}/*_align.sh ; do 
  qsub ${align}
done

```


```{bash eval=FALSE}
FQ1=../data/SRR1608907_1.fastq.gz
FQ2=../data/SRR1608907_2.fastq.gz
star_genome100=path_to_genome_directory/star_index_hg19_99

Rscript ../functions/run_STAR.R --genome_index $star_index_hg19_99 \
--fastqfiles $FQ1,$FQ2 \
--sampleName SRR1608907 \
--outdir ../results/star_1pass \
--STARmode "1Pass" 
```

The `R` function above is a wrapper for the `STAR` call below:

```{bash eval=FALSE}
# Version STAR/2.5

STAR --genomeDir path_to_star_index_hg19 \
--readFilesIn $FQ1 $FQ2 --runThreadN 27 --chimSegmentMin 10 --readFilesCommand zcat --alignSJoverhangMin 8 --outBAMcompression 10 --alignSJDBoverhangMin 1 --limitBAMsortRAM 85741557872 --outFilterMismatchNmax 999 --alignIntronMin 20 --alignIntronMax 200000 --alignMatesGapMax 20000 --outFileNamePrefix path_to_star_1pass/SampleName --outSAMtype BAM SortedByCoordinate --outFilterType BySJout --outFilterMultimapNmax 15
```

To see all the arguments available:

```{bash eval=FALSE}
Rscript ../functions/run_STAR.R --help
```

After running STAR on all the fastq files available we can collect all the splice junctions from the first pass and use them for the second pass.

```{bash eval=FALSE}
# concatenate splice junctions from all samples from ran in pass1
cat ../results/star_1pass/*SJ.out.tab > ../results/star_1pass/combined_sj.out.tab
# Dobin suggests to remove chrm cause they are usually False positives
awk '!/chrM/' ../results/star_1pass/combined_sj.out.tab > ../results/star_1pass/combined_sj_nochrM.out.tab
```

Again, for quality check, have a look at the amazing alignment summary enabled by [MultiQC](http://multiqc.info/).

```{bash eval=FALSE}
multiqc ../results/star_1pass --interactive -n "STAR_1passQC" -o ../results
```


## STAR-2pass

The second pass alignment is exactly the same as the first one with only a few differences:

- the *sjfile* input created combining the splice junctions from the first pass
- STAR is run with the option of output chimeric reads switched on. This will allow fusion analysis.

The ouput of STAR is a bamfile already sorted by coordinate with the suffix `Aligned.sortedByCoord.out.bam`. At this stage we can also run two more steps `post_align_qc1.sh` and `post_align_qc2.sh`, discussed below.

```{bash eval=FALSE}
FQ1=../data/SRR1608907_1.fastq.gz
FQ2=../data/SRR1608907_2.fastq.gz
star_genome100=path_to_genome_directory/star_index_hg19_99

Rscript ../functions/run_STAR.R \
--genome_index $star_index_hg19_99 \
--fastqfiles $FQ1,$FQ2 \
--sampleName SRR1608907 \
--outdir ../results/star_2pass --STARmode "2PassMulti" \
--sjfile ../results/star_1pass/combined_sj_nochrM.out.tab

# Run featurecounts and collect fragment sizes for QC
../functions/post_align_qc1.sh \
path_to_genome.fa \
path_to_genes.gtf \
../results/star_2pass/SRR1608907.Aligned.sortedByCoord.out.bam \
SRR1608907  # sample name

# Pre-process bamfile (add Read groups etc..)
../functions/post_align_qc2.sh \ ../results/star_2pass/SRR1608907.Aligned.sortedByCoord.out.bam \
SRR1608907 \
path_to_genome.fa \
SRR1608907 

```

## Details about post-alignment functions {#sec:post-align}

* `post_align_qc1.sh` is optional:

  1. Runs [featureCounts](http://bioinf.wehi.edu.au/featureCounts/) to get gene counts and compute PCA to evaulate the concordance between bamfiles sequenced on different lanes. This allows a QC before merging the different bamfiles into a single one.
  2. Runs [CollectMultipleMetrics](https://broadinstitute.github.io/picard/command-line-overview.html) to collect the fragment distribution of the bamfiles (only possible with PE reads). This is also a good QC to check that the fragment distribution of bamfiles on different lanes is the same. 
  
* `post_align_qc2.sh` contains necessary pre-prcessing steps:

  1. Marks PCR duplicates (using [sambamba markdup](http://lomereiter.github.io/sambamba/docs/sambamba-markdup.html))
  2. Add Read Groups to single runs before merging bamfiles (using [AddOrReplaceReadGroups](https://broadinstitute.github.io/picard/command-line-overview.html)). Even if files do not need to be merges, `GATK` requires read groups to be added bamfiles.
  3. Run [ValidateSamFile](https://broadinstitute.github.io/picard/command-line-overview.html) to check for errors in the final bamfile.

In order, its arguments are:
  
  1. Path to **aligned bamfile**;
  2. `SampleName`. This is the name of the sample applied to the `RGID` and `RGPU` fields below.
  4. `SampleName of the run`. If a sample was sequenced across different lanes you need to set lane-specific read groups to each separate bamfile (e.g. SampleName_L1, SampleName_L2). This sample name will be used for the fields `RGLB` and `RGSM` in the `AddOrReplaceReadGroups` groups below. See Chapter \@(ch:merge-bamfiles) for merging bamfiles.

```{bash eval=FALSE}
# Picard tool function to add read groups to a bamfile
AddOrReplaceReadGroups \
		I= ./star_2pass/SRR1608907.Aligned.sortedByCoord.out.bam \
		O= ./star_2pass/SRR1608907.Aligned.sortedByCoord.out.RG.bam \
		RGID=SRR1608907 \
		RGPU=SRR1608907 \
		RGLB=SRR1608907_L1 \
		RGPL="illumina" \
		RGSM=SRR1608907_L1
```

After running `post_align_qc2.sh` a file with the suffix `Aligned.reorderedDupl.rg.bam` will be created where read groups are added and PCR duplicated reads marked.

```{bash eval=FALSE,echo=FALSE}
# save list of fastq files into a variable
samplein_runs=$(find ${fastqdir} -maxdepth 1 -name "*1.fastq*" | sort)

for fastqin in $samplein_runs ; do

  FQ1=$fastqin
  FQ2=${FQ1/_1/_2}
  
  bamout=$(echo $(basename $fastqin) | cut -f1 -d "_") # to be personalised according to file name

  echo '#!/bin/bash' > ${star_2pass}/${bamout}_align.sh
  echo "#PBS -q submit" >> ${star_2pass}/${bamout}_align.sh
  echo "#PBS -l nodes=1:ppn=1,mem=1gb" >> ${star_2pass}/${bamout}_align.sh
  echo "#PBS -N ${bamout}" >> ${star_2pass}/${bamout}_align.sh
  echo "#PBS -o ${bamout}_out" >> ${star_2pass}/${bamout}_align.sh
  echo "#PBS -e ${bamout}_err" >> ${star_2pass}/${bamout}_align.sh
  echo  "" >> ${star_2pass}/${bamout}_align.sh
  echo 'module add STAR' >> ${star_2pass}/${bamout}_align.sh
  echo 'module load sambamba' >> ${script_dir}/${bamout}_align.sh
  echo 'module load picard-tools' >> ${script_dir}/${bamout}_align.sh
  echo 'module load R' >> ${script_dir}/${bamout}_align.sh

  echo  Rscript ${workdir}/run_STAR.R --genome_index ${star_genome100} --fastqfiles $FQ1,$FQ2 \
    --sampleName ${bamout} --outdir ${star_2pass} --STARmode "2PassMulti" \
    --sjfile ${star_2pass}/combined_sj_nochrM.out.tab >> ${star_2pass}/${bamout}_align.sh

  ######################################
  # Mark duplicates, addRG and Validate 
  ######################################
  bamin=${star_2pass}/${bamout}Aligned.sortedByCoord.out.bam
  
  echo ${workdir}/post_align_qc1.sh ${genome_fasta} ${gtf} ${bamin} ${bamout} >> ${star_2pass}/${bamout}_align.sh
  
  echo ${workdir}/post_align_qc2.sh ${bamin} ${bamout} ${genome_fasta} ${bamout} >> ${star_2pass}/${bamout}_align.sh


done



```

This time *MultiQC* will give us a summary output also of the fragment distributions and gene counts if the output files are stored within the `star_2pass` folder.

```{bash eval=FALSE}
multiqc ../results/star_2pass --interactive -n "STAR_2passQC" -o ../results
```
